File: data-provider/src/main/scala/velocorner/api/weather/Weather.scala
package velocorner.api.weather

import org.joda.time.DateTime
import play.api.libs.json._
import velocorner.model.EpochFormatter

/**
  * https://openweathermap.org/forecast5#format
  */
object WindDescription {
  implicit val responseFormat = Format[WindDescription](Json.reads[WindDescription], Json.writes[WindDescription])
}

case class WindDescription(
                            speed: Double, // m/s
                            deg: Double    // degrees
                          )

object CloudDescription {
  implicit val responseFormat = Format[CloudDescription](Json.reads[CloudDescription], Json.writes[CloudDescription])
}

case class CloudDescription(
                             all: Int // %
                           )

object RainDescription {
  implicit val responseFormat = Format[RainDescription](Json.reads[RainDescription], Json.writes[RainDescription])
}

case class RainDescription(
                            `3h`: Option[Double] // mm
                          )

object SnowDescription {
  implicit val responseFormat = Format[SnowDescription](Json.reads[SnowDescription], Json.writes[SnowDescription])
}

case class SnowDescription(
                            `3h`: Option[Double] // volume
                          )

object WeatherDescription {
  implicit val responseFormat = Format[WeatherDescription](Json.reads[WeatherDescription], Json.writes[WeatherDescription])
}

case class WeatherDescription(
                               id: Long, // 800 - weather condition code
                               main: String, // "Clear"
                               description: String, // "clear sky"
                               icon: String // weather icon 04d
                             )

object WeatherInfo {
  implicit val responseFormat = Format[WeatherInfo](Json.reads[WeatherInfo], Json.writes[WeatherInfo])
}

case class WeatherInfo(
                        temp: Float, // C
                        temp_min: Float, // C
                        temp_max: Float, // C
                        humidity: Float, // %
                        pressure: Float // hPa
                      )

object Weather {

  implicit val dateTimeFormat = EpochFormatter.create
  val writes = new Writes[Weather] {
    override def writes(o: Weather): JsValue = {
      val baseJs: JsObject = Json.writes[Weather].writes(o).as[JsObject]
      val typeJs: JsString = Writes.StringWrites.writes("Weather")
      JsObject(baseJs.fields :+ ("type" -> typeJs))
    }
  }
  implicit val weatherFormat = Format[Weather](Json.reads[Weather], writes)
}

// one of the entry points, the response contains a list of Weather structures, also stored in database
case class Weather(
                    dt: DateTime,
                    main: WeatherInfo,
                    weather: List[WeatherDescription],
                    snow: Option[SnowDescription],
                    rain: Option[RainDescription],
                    clouds: CloudDescription,
                    wind: WindDescription
                  )

// the list of Weather structures is associated with a city as well
object City {
  implicit val responseFormat = Format[City](Json.reads[City], Json.writes[City])
}

case class City(
                 id: Long,
                 name: String, // plain name, such as Zurich, Budapest
                 country: String // ISO code 2 letter
               )

File: src/main/kotlin/velocorner/weather/model/Weather.kt
package velocorner.weather.model

import kotlinx.serialization.KSerializer
import kotlinx.serialization.Serializable
import kotlinx.serialization.Serializer
import kotlinx.serialization.descriptors.PrimitiveKind
import kotlinx.serialization.descriptors.PrimitiveSerialDescriptor
import kotlinx.serialization.descriptors.SerialDescriptor
import kotlinx.serialization.encoding.Decoder
import kotlinx.serialization.encoding.Encoder
import java.time.Instant
import java.time.OffsetDateTime
import java.time.ZoneOffset

/**
 * The response structures from openweathermap.
 * https://openweathermap.org/current
 */

/**
 * Response example.
 * <code><pre>
 * {
 *   "coord": {
 *     "lon": 8.52,
 *     "lat": 47.31
 *   },
 *   "weather": [
 *   {
 *     "id": 801,
 *     "main": "Clouds",
 *     "description": "few clouds",
 *     "icon": "02d"
 *   }
 *   ],
 *   "base": "stations",
 *   "main": {
 *     "temp": 7.57,
 *     "feels_like": 4.53,
 *     "temp_min": 5,
 *     "temp_max": 10,
 *     "pressure": 1012,
 *     "humidity": 81
 *   },
 *   "visibility": 10000,
 *   "wind": {
 *     "speed": 2.6,
 *     "deg": 20
 *   },
 *   "clouds": {
 *     "all": 20
 *   },
 *   "dt": 1602670919,
 *   "sys": {
 *     "type": 1,
 *     "id": 6932,
 *     "country": "CH",
 *     "sunrise": 1602654185,
 *     "sunset": 1602693616
 *   },
 *   "timezone": 7200,
 *   "id": 2661861,
 *   "name": "Adliswil",
 *   "cod": 200
 * }
 * </pre></code>
 */

@Serializable
data class WeatherDescription(
    val id: Int, // 800 - weather condition code
    val main: String, // "Clear"
    val description: String, // "clear sky"
    val icon: String // weather icon 04d
)

@Serializable
data class WeatherInfo(
    val temp: Float, // C
    val temp_min: Float, // C
    val temp_max: Float, // C
    val humidity: Float, // %
    val pressure: Float // hPa
)

@Serializer(forClass = OffsetDateTime::class)
object OffsetDateTimeSerializer : KSerializer<OffsetDateTime> {

    override val descriptor: SerialDescriptor = PrimitiveSerialDescriptor("WithCustomDefault", PrimitiveKind.LONG)

    override fun serialize(encoder: Encoder, value: OffsetDateTime) {
        encoder.encodeLong(value.toInstant().toEpochMilli() / 1000)
    }

    override fun deserialize(decoder: Decoder): OffsetDateTime {
        val epoch = decoder.decodeLong()
        return OffsetDateTime.ofInstant(Instant.ofEpochSecond(epoch), ZoneOffset.UTC)
    }
}

@Serializable
data class SunriseSunsetInfo(
    @Serializable(with = OffsetDateTimeSerializer::class)
    val sunrise: OffsetDateTime,
    @Serializable(with = OffsetDateTimeSerializer::class)
    val sunset: OffsetDateTime
)

@Serializable
data class Coord(val lon: Double, val lat: Double)

@Serializable
data class SnowDescription(
    val `3h`: Double? = null // volume
)

@Serializable
data class RainDescription(
    val `3h`: Double? = null // mm
)

@Serializable
data class CloudDescription(
    val all: Int // %
)

@Serializable
data class WindDescription(
    val speed: Double, // m/s
    val deg: Double // degrees
)

@Serializable
data class City(
    val id: Long,
    val name: String, // plain name, such as Zurich, Budapest
    val country: String // ISO code 2 letter
)

// one of the entry points, the response contains a list of Weather structures, also stored in database
@Serializable
data class Weather(
    @Serializable(with = OffsetDateTimeSerializer::class)
    val dt: OffsetDateTime,
    val main: WeatherInfo,
    val weather: List<WeatherDescription>,
    val snow: SnowDescription? = null, // default value needed to signal optionality
    val rain: RainDescription? = null,
    val clouds: CloudDescription,
    val wind: WindDescription
)

// response expected from OpenWeatherMap API
@Serializable
data class CurrentWeatherResponse(
    val cod: Int,
    val weather: List<WeatherDescription>? = null,
    val main: WeatherInfo? = null,
    val sys: SunriseSunsetInfo? = null,
    val coord: Coord? = null,
    @Serializable(with = OffsetDateTimeSerializer::class)
    val dt: OffsetDateTime? = null
)

// response of the service API and persisted in the caching layer (converted from CurrentWeatherResponse)
@Serializable
data class CurrentWeather(
    val location: String, // city[, country iso 2 letters]
    @Serializable(with = OffsetDateTimeSerializer::class)
    val timestamp: OffsetDateTime,
    val bootstrapIcon: String, // bootstrap icon derived from the current weather code
    val current: WeatherDescription,
    val info: WeatherInfo,
    val sunriseSunset: SunriseSunsetInfo,
    val coord: Coord
)

// response expected from OpenWeatherMap API
@Serializable
data class ForecastWeatherResponse(
    val cod: String,
    val list: List<Weather>? = null,
    val city: City? = null
)

// response of the service API and persisted in the caching layer (converted from ForecastWeatherResponse)
// https://openweathermap.org/forecast5#format
@Serializable
data class ForecastWeather(
    val location: String, // city[, country iso 2 letters]
    @Serializable(with = OffsetDateTimeSerializer::class)
    val timestamp: OffsetDateTime,
    val forecast: Weather
)
